pipelines/controller/src/main/java/com/google/fhir/analytics/FlinkConfiguration.java

```diff
@@ -0,0 +1,167 @@
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import org.apache.flink.configuration.ConfigConstants;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.DeploymentOptions;
+import org.apache.flink.configuration.GlobalConfiguration;
+import org.apache.flink.configuration.IllegalConfigurationException;
+import org.apache.flink.configuration.MemorySize;
+import org.apache.flink.configuration.TaskManagerOptions;
+import org.apache.flink.runtime.taskexecutor.TaskExecutorResourceUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.FileSystemUtils;
+
+/** This class is responsible for the generation and validation of Flink configuration file. */
+public class FlinkConfiguration {
+
+  private static final Logger logger = LoggerFactory.getLogger(FlinkConfiguration.class.getName());
+  static final String TEMP_FLINK_CONF_DIR = "tmp-flink-conf";
+
+  // This value is arrived based on the maximum number of reshuffle/partition operations that can
+  // occur in the pipelines. This has to be updated if any modifications are made to the pipelines
+  // that impact the reshuffle/partition operations.
+  static final int NUMBER_OF_RESHUFFLES = 8;
+  private static final int DEFAULT_PARALLELISM = Runtime.getRuntime().availableProcessors();
+  static final String DEFAULT_MANAGED_MEMORY_SIZE = "256mb";
+  private static final String KEY_VALUE_FORMAT = "%s: %s";
+  private String flinkConfDir;
+
+  public String getFlinkConfDir() {
+    return flinkConfDir;
+  }
+
+  /**
+   * This method auto-generates/validates the flink-conf.yaml file. The auto generated file contains
+   * parameter values which are optimised for the pipelines to run without fail. In case of
+   * auto-generated file, the file is placed in a temporary directory and will be deleted once the
+   * application is stopped.
+   *
+   * <p>In particular the task manager network buffer size for each pipeline is inspired from this
+   * <a
+   * href="https://nightlies.apache.org/flink/flink-docs-release-1.8/ops/config.html#configuring-the-network-buffers">link</a>
+   *
+   * <p>This method also fails fast if the generated/passed flink configuration values might result
+   * in failure based on the JVM Heap memory values.
+   *
+   * @param dataProperties the application properties
+   * @throws IOException
+   */
+  void initialiseFlinkConfiguration(DataProperties dataProperties) throws IOException {
+    if (dataProperties.isAutoGenerateFlinkConfiguration()) {
+      Path confPath = Files.createTempDirectory(TEMP_FLINK_CONF_DIR);
+      logger.info("Creating Flink temporary configuration directory at {}", confPath);
+      Path filePath = Paths.get(confPath.toString(), GlobalConfiguration.FLINK_CONF_FILENAME);
+      validateAndCreateConfFile(filePath, dataProperties);
```

Comment:
The configuration file that gets created in Temp folder or the one that gets set by the env. variable FLINK_CONF_DIR gets passed to the Flink Execution Environment as per this [method](https://github.com/apache/beam/blob/444e10d29e04e3e144ef08c0a505bd76aa08984f/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java#L87).

However for non-local modes there are many configuration parameters that could be used, so I think we should recommend the users to pass their own custom file instead of Auto generated one.

Reply:
Can you please document it somewhere, i.e., in one of the README files or maybe one of the Wiki pages? I think it is fine for now as I have not seen a user with a Flink cluster (so Flink is usually used locally); but we should leave that option open (horizontal scalability).

Reply:
Added checks in the code to invoke this code if it is local and autoGenerate is enabled. In case of non-local mode this will be by-passed. Documented in the `pipelines/controller/apllication.yaml` file as well